### PR TITLE
make get_browser_version_from_os  faster

### PR DESCRIPTION
**Pull request by np-8**
- Fixes https://github.com/SergeyPirogov/webdriver_manager/issues/323

### CODE REVIEW

File: webdriver_manager/utils.py [hunk 1]
```diff
@@ -153,6 +153,84 @@ def windows_browser_apps_to_cmd(*apps: str) -> str:
...
+        OSType.MAC: r'/Applications/Firefox.app/Contents/MacOS/firefox --version',
+        OSType.WIN: windows_browser_apps_to_cmd(
+            r'(Get-Item -Path "$env:PROGRAMFILES\Mozilla Firefox\firefox.exe").VersionInfo.FileVersion',
+            r'(Get-Item -Path "$env:PROGRAMFILES (x86)\Mozilla Firefox\firefox.exe").VersionInfo.FileVersion',
+            r"(Get-Item (Get-ItemProperty 'HKLM:\SOFTWARE\Microsoft\Windows\CurrentVersion\App Paths\firefox.exe').'(Default)').VersionInfo.ProductVersion",
+            r'(Get-ItemProperty -Path Registry::"HKLM\SOFTWARE\Mozilla\Mozilla Firefox").CurrentVersion'
+        ),
+    },
+}
+
```
Comment (aleksandr-kotlyar) [inline webdriver_manager/utils.py:233]: There is 1 blank line. Should be 2 blank lines between constants/functions/classes in module. Add 1 more blank line here.

**Review by aleksandr-kotlyar — CHANGES_REQUESTED** [review]
@np-8 nice optimization. Before merge please take an attention to requested changes.

**Comment by np-8** [comment]
Thanks! 

I created the changes. One additional optimization which could be used is to put the `CMD_MAPPING` inside a function, like `get_cmd_mapping`. Currently, as `CMD_MAPPING` is global variable in the module, if you import anything from  `webdriver_manager.utils`, the `CMD_MAPPING` is created (which takes some time). Sometimes pattern like this is used:

```python

def get_cmd_mapping():
    if not hasattr(get_cmd_mapping, 'mapping'):
         # calculate the mapping
         get_cmd_mapping.mapping = mapping 
    return get_cmd_mapping.mapping    
```

in this way the `CMD_MAPPING` is only calculated once and only when it is accessed (and not when the module is imported).

File: webdriver_manager/utils.py [hunk 2]
```diff
@@ -120,6 +120,18 @@ def get_filename_from_response(response, name):
 
     return filename
 
+def determine_powershell():
```
Comment (aleksandr-kotlyar) [inline webdriver_manager/utils.py:123]: Take a look at 
[Line 123:1](https://github.com/SergeyPirogov/webdriver_manager/blob/f905d45ca796775ee20ddb6be77580edf6029667/webdriver_manager/utils.py#L123): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

**Review by aleksandr-kotlyar — CHANGES_REQUESTED** [review]

**Comment by aleksandr-kotlyar** [comment]
@np-8 You thought about optimization but you have been faced the problem each import of `utils` calls `CMD_MAPPING` initialization. That doesn't seem like optimization then.  This PR will not truly optimize webdriver_manager for now.


But you are talking about the new workaround in this message
> I created the changes. One additional optimization which could be used is to put the `CMD_MAPPING` inside a function, like `get_cmd_mapping`. Currently, as `CMD_MAPPING` is global variable in the module, if you import anything from `webdriver_manager.utils`, the `CMD_MAPPING` is created (which takes some time). Sometimes pattern like this is used:
> 
> ```python
> def get_cmd_mapping():
>     if not hasattr(get_cmd_mapping, 'mapping'):
>          # calculate the mapping
>          get_cmd_mapping.mapping = mapping 
>     return get_cmd_mapping.mapping    
> ```
> 
> in this way the `CMD_MAPPING` is only calculated once and only when it is accessed (and not when the module is imported).

@np-8 Do you think that new workaround will work good in all cases?

**Comment by np-8** [comment]
@aleksandr-kotlyar I decided to go even further and to make only commands for the `browser_type` and `os_type` when requested, and used caching in `_CMD_MAPPING`.

- If entry of `_CMD_MAPPING` is a string it is assumed to be the final form of the command
- If entry of `_CMD_MAPPING` is not a string, then the respective function (from `CMD_MAPPING_FUNCS`) is called to create the final command and save it into `_CMD_MAPPING`.

The commands are not created on `import webdriver_manager.utils`, and only the commands that are needed are created.

Of course, one could argue that the commands could be saved as strings for every browser type and os and hard-coded to webdriver_manager package. That's one option, too. I guess that has not been done because of some readability/maintainability reason.

**Comment by np-8** [comment]
### Timing comparisons (this PR)

Here are some timing comparisons with & without the fix, using

```
%timeit get_browser_version_from_os('google-chrome')
```
With 9d2f12f (this PR), first and second run

952 ms ± 37.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
670 ms ± 14.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

with 9e03223 (without this PR),  first and second run

1.55 s ± 37 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
1.56 s ± 86.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


### Looking at the root cause

I was still wondering how could concatenating few strings can take such a large time. Probably the reason is actually repeated calls to `determine_powershell` 

```
In [5]: %timeit determine_powershell()
209 ms ± 15 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Having separate function for `get_command_to_read_browser_version(browser_type, os_name)` might be good reason regardless of what is the underlying reason, as it tidies `get_browser_version_from_os`, and helps to separate things.


### winreg-based solution 

As on Windows, one could read registry values directly, if they can be trusted, it might be faster to read those directly from registry rather than using powershell as intermediate step. I took helper functions from [this SO answer](https://stackoverflow.com/a/40583388/3015186) to creat function that just reads the Google chrome version from the registry. Here are the timings:

```
[WDM] - Current google-chrome version is UNKNOWN
715 ms ± 37.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [3]: %timeit get_chrome_version()
32.6 µs ± 873 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
````

As one can see, the **reading the version from the registry is about 22.000 times faster** that spawning powershell process to do the same thing. More importantly, the read chrome version will be correct (and not `UNKNOWN`):

```
In [4]: get_browser_version_from_os('google-chrome')
[WDM] - Could not get version for google-chrome with the command:  powershell -EncodedCommand JABFAHIAcgBvAHIAQQBjAHQAaQBvAG4AUAByAGUAZgBlAHIAZQBuAGMAZQA9ACcAcwBpAGwAZQBuAHQAbAB5AGMAbwBuAHQAaQBuAHUAZQAnACAAOwAgACgAJwAoAEcAZQB0AC0ASQB0AGUAbQAgAC0AUABhAHQAaAAgACIAJABlAG4AdgA6AFAAUgBPAEcAUgBBAE0ARgBJAEwARQBTAFwAXABHAG8AbwBnAGwAZQBcAFwAQwBoAHIAbwBtAGUAXABcAEEAcABwAGwAaQBjAGEAdABpAG8AbgBcAFwAYwBoAHIAbwBtAGUALgBlAHgAZQAiACkALgBWAGUAcgBzAGkAbwBuAEkAbgBmAG8ALgBGAGkAbABlAFYAZQByAHMAaQBvAG4AJwAsACAAJwAoAEcAZQB0AC0ASQB0AGUAbQAgAC0AUABhAHQAaAAgACIAJABlAG4AdgA6AFAAUgBPAEcAUgBBAE0ARgBJAEwARQBTACAAKAB4ADgANgApAFwAXABHAG8AbwBnAGwAZQBcAFwAQwBoAHIAbwBtAGUAXABcAEEAcABwAGwAaQBjAGEAdABpAG8AbgBcAFwAYwBoAHIAbwBtAGUALgBlAHgAZQAiACkALgBWAGUAcgBzAGkAbwBuAEkAbgBmAG8ALgBGAGkAbABlAFYAZQByAHMAaQBvAG4AJwAsACAAJwAoAEcAZQB0AC0ASQB0AGUAbQAgAC0AUABhAHQAaAAgACIAJABlAG4AdgA6AEwATwBDAEEATABBAFAAUABEAEEAVABBAFwAXABHAG8AbwBnAGwAZQBcAFwAQwBoAHIAbwBtAGUAXABcAEEAcABwAGwAaQBjAGEAdABpAG8AbgBcAFwAYwBoAHIAbwBtAGUALgBlAHgAZQAiACkALgBWAGUAcgBzAGkAbwBuAEkAbgBmAG8ALgBGAGkAbABlAFYAZQByAHMAaQBvAG4AJwAsACAAJwAoAEcAZQB0AC0ASQB0AGUAbQBQAHIAbwBwAGUAcgB0AHkAIAAtAFAAYQB0AGgAIABSAGUAZwBpAHMAdAByAHkAOgA6ACIASABLAEMAVQBcAFwAUwBPAEYAVABXAEEAUgBFAFwAXABHAG8AbwBnAGwAZQBcAFwAQwBoAHIAbwBtAGUAXABcAEIATABCAGUAYQBjAG8AbgAiACkALgB2AGUAcgBzAGkAbwBuACcALAAgACcAKABHAGUAdAAtAEkAdABlAG0AUAByAG8AcABlAHIAdAB5ACAALQBQAGEAdABoACAAUgBlAGcAaQBzAHQAcgB5ADoAOgAiAEgASwBMAE0AXABcAFMATwBGAFQAVwBBAFIARQBcAFwAVwBvAHcANgA0ADMAMgBOAG8AZABlAFwAXABNAGkAYwByAG8AcwBvAGYAdABcAFwAVwBpAG4AZABvAHcAcwBcAFwAQwB1AHIAcgBlAG4AdABWAGUAcgBzAGkAbwBuAFwAXABVAG4AaQBuAHMAdABhAGwAbABcAFwARwBvAG8AZwBsAGUAIABDAGgAcgBvAG0AZQAiACkALgB2AGUAcgBzAGkAbwBuACcAKQAgADsA
[WDM] - Current google-chrome version is UNKNOWN
Out[4]: 'UNKNOWN'

In [5]: get_chrome_version()
Out[5]: '98.0.4758.102'
```

I put a quick-and-dirty proof of concept to [fast-version-read-winreg](https://github.com/np-8/webdriver_manager/tree/fast-version-read-winreg) branch (ba181f9). See the `webdriver_manager\winutils.py:get_chrome_version()`.

**Pull request closed without merging**